src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,505 @@
+    };
+
+    let ret = store.into_data().buffer;
+
+    Some(GlyphBuffer(ret))
+}
+
+// ===========
+// functions imported  into WASM
+// ===========
+
+// fn face_get_upem(face: u32) -> u32;
+// Returns the units-per-em of the font face.
+fn face_get_upem(caller: Caller<'_, ShapingData>, _face: u32) -> u32 {
+    caller.data().font.units_per_em as u32
+}
+
+// fn font_get_face(font: u32) -> u32;
+// Creates a new face token from the given font token.
+fn font_get_face(caller: Caller<'_, ShapingData>, _font: u32) -> u32 {
+    // From HarfBuzz docs:
+    // (In the following functions, a font is a specific instantiation of a face at a
+    // particular scale factor and variation position.)
+    //
+    // I am unsure how to represent that in rustybuzz.
+    // er .. do what here?
+    0
+}
+
+// fn font_get_glyph(font: u32, unicode: u32, uvs: u32) -> u32;
+// Returns the nominal glyph ID for the given codepoint, using the cmap table of the font to map Unicode codepoint (and variation selector) to glyph ID.
+fn font_get_glyph(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    codepoint: u32,
+    uvs: u32,
+) -> u32 {
+    let Some(codepoint) = char::from_u32(codepoint) else {
+        return 0;
+    };
+
+    if uvs == 0 {
+        caller.data().font.glyph_index(codepoint).unwrap().0 as u32
+    } else {
+        let Some(uvs) = char::from_u32(uvs) else {
+            return 0;
+        };
+        caller
+            .data()
+            .font
+            .glyph_variation_index(codepoint, uvs)
+            .unwrap_or_default()
+            .0 as u32
+    }
+}
+
+// fn font_get_scale(font: u32, x_scale: *mut i32, y_scale: *mut i32);
+// Returns the scale of the current font.
+// Just return the upem as rustybuzz has no scale.
+fn font_get_scale(mut caller: Caller<'_, ShapingData>, _font: u32, x_scale: u32, y_scale: u32) {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+    let upem = caller.data().font.units_per_em();
+
+    memory.write(
+        &mut caller.as_context_mut(),
+        x_scale as usize,
+        &upem.to_le_bytes(),
+    );
+    memory.write(
+        &mut caller.as_context_mut(),
+        y_scale as usize,
+        &upem.to_le_bytes(),
+    );
+}
+
+// fn font_get_glyph_extents(font: u32, glyph: u32, extents: *mut CGlyphExtents) -> bool;
+// Returns the glyph's extents for the given glyph ID at current scale and variation settings.
+fn font_get_glyph_extents(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    extents: u32,
+) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut glyph_extents = hb_glyph_extents_t::default();
+    let ret = caller
+        .data()
+        .font
+        .glyph_extents(GlyphId(glyph as u16), &mut glyph_extents);
+    if ret {
+        memory.write(
+            caller.as_context_mut(),
+            extents as usize,
+            bytemuck::bytes_of(&glyph_extents),
```

Comment:
won't this break on big endian machines?

Reply:
No idea.

Reply:
We do not support big endian machines to begin with.

Reply:
Is that mentioned somewhere? And are there existing code paths that don't work on big endian?

Reply:
I don't think it's mentioned. All I know is that we do not test for big endian architectures. Mainly because they are so rare.
It might work, might not.